### PR TITLE
SFT Script with HuggingFace chat template support

### PR DESCRIPTION
**Pull request by jenchen13**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Support using the HuggingFace tokenizer `apply_chat_template` during SFT with chat-style datasets, and update distillation script to support finetuning for chat datasets as well as SFT distillation.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# TODO

- [ ] update [documentation](https://gitlab-master.nvidia.com/nemo-framework-tme/documentation/-/blob/main/docs/user-guide/model-optimization/distillation/distillation.rst?ref_type=heads)

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: scripts/llm/finetuning.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        micro_batch_size=args.micro_batch_size,
+        use_hf_tokenizer_chat_template=True,
+    )
+
+    executor: run.Executor
+
+    if args.slurm:
+        # TODO: Set your custom parameters for the Slurm Executor.
+        executor = slurm_executor(
+            user="jennifchen",
```
Comment (jenchen13) [inline scripts/llm/finetuning.py:272]: TODO remove these before merge

File: nemo/collections/llm/gpt/data/core.py [hunk 2]
```diff
@@ -913,17 +927,24 @@ def _process_example(self, example):
...
+        else:
+            if "conversations" in example:  # convert ShareGPT format to HuggingFace chat template format
+                example = _transform_to_chat_message(example)
+            result = _preprocess_hf_chat_template(
+                example,
+                self.tokenizer,
+            )
         # store metadata in dataset, in case user may have keys required in the prediction json files
-        metadata = {k: v for k, v in example.items() if k not in ['conversations']}
+        metadata = {k: v for k, v in example.items() if k not in {'conversations', 'messages'}}
```
Comment (AAnoosheh) [inline nemo/collections/llm/gpt/data/core.py:947]: nit: Sets less readable than list here (and actually slower overall when so few elements inside it :P)

File: scripts/llm/finetuning.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+def get_parser():
+    parser = argparse.ArgumentParser(description="NeMo2.0 Pretraining")
+    parser.add_argument(
+        "--recipe",
+        type=str,
+        default="llama3_8b",
+        help="Choose NeMo 2.0 recipe. Recipes are named in the format of <model_name>_<model_size>(_<long_sequence_length> or other special settings)",
+    )
+    parser.add_argument(
+        "--resume_path",
```
Comment (AAnoosheh) [inline scripts/llm/finetuning.py:54]: inconsistent underscore use

File: nemo/collections/llm/gpt/data/chat.py [hunk 4]
```diff
@@ -26,6 +26,26 @@ class ChatDataModule(FineTuningDataModule):
     See base class `FineTuningDataModule` for more details.
     """
 
+    def __init__(
+        self,
+        use_hf_tokenizer_chat_template: bool = False,
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply nemo/collections/llm/gpt/data/chat.py:31]: ah true, should I just copy over the parent class's constructor? it seems verbose but probably safest

[the remark this answers is not in the document]
Comment (jenchen13) [reply nemo/collections/llm/gpt/data/chat.py:50]: updated both constructors

File: scripts/llm/finetuning.py
Comment (cuichenx) [inline scripts/llm/finetuning.py:1]: scripts are not the recommended entry point in nemo2 (we want to encourage the use of nemo run /cli). Is there a specific reason to add this script? we would want to update the docs rather than providing a script

File: nemo/collections/llm/api.py [hunk 5]
```diff
@@ -203,6 +204,8 @@ def finetune(
         optim (Optional[OptimizerModule]): The optimizer module to be used. If not provided, the default
             optimizer from the model will be used.
         peft (Optional[PEFT]): A PEFT (Parameter-Efficient Fine-Tuning) configuration to be applied.
+        tokenizer (Optional[TokenizerType]): Tokenizer setting to be applied. Can be 'data' or 'model'
+            or an instance of TokenizerSpec.
```
Comment (cuichenx) [inline nemo/collections/llm/api.py:208]: perhaps you can explain the reason to specify a new tokenizer rather than using the one that comes with the model here in the docstring

File: scripts/llm/finetuning.py
Comment (jenchen13) [reply scripts/llm/finetuning.py:1]: we just wanted to provide an easy way for user to run SFT by providing a recipe name + data path without having to write custom code. Currently to do SFT on a non-SquAD dataset you need to write NeMo code because the pretrain/finetune [recipes](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/recipes/llama3_8b.py#L244) hardcodes the dataset (in finetune it's SquAD). Would you suggest changing the `finetune_recipe` API to add additional args for dataset path, use HF chat template etc? My main concern with that is I'd have to change every single recipe file ...

Comment (jenchen13) [reply scripts/llm/finetuning.py:1]: Chenhan and I discussed, and for now we decided to update the existing `gpt_distillation.py` script to support SFT as well (now renamed to `gpt_sft_distillation.py`) instead of adding a completely new script for SFT. 

Our concern with modifying the NeMo CLI is that it would require a very significant API change, and also that the current NeMo recipes are very tightly coupled with NeMo Run. We can discuss more later about medium/long term API changes.

File: nemo/collections/llm/gpt/data/utils.py [hunk 6]
```diff
@@ -802,6 +802,50 @@ def _preprocess(
     return dict(input_ids=input_ids, mask=mask, context_ids=context_ids, answer_ids=answer_ids)
 
 
+def _transform_to_chat_message(source: dict):
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply nemo/collections/llm/gpt/data/utils.py:805]: it's probably `dict[str, list[dict[str, str]]]` but I'll just do `dict[str, list]` ?

File: nemo/collections/llm/api.py [hunk 7]
```diff
@@ -203,6 +204,9 @@ def finetune(
         optim (Optional[OptimizerModule]): The optimizer module to be used. If not provided, the default
             optimizer from the model will be used.
         peft (Optional[PEFT]): A PEFT (Parameter-Efficient Fine-Tuning) configuration to be applied.
+        tokenizer (Optional[TokenizerType]): Tokenizer setting to be applied. Can be 'data' or 'model'
+            or an instance of TokenizerSpec. If 'data' uses the data loader's tokenizer (useful if data loader
+            uses a separate tokenizer from model).
```
Comment (cuichenx) [inline nemo/collections/llm/api.py:209]: Sorry, i meant something like 
"Specifying a new tokenizer (rather than the tokenizer from the model checkpoint) is useful for expanding vocabulary or adding special tokens"
If there is something else, feel free to add it too. I just thought it may be not straightforward to some users why we don't use the same tokenizer that comes from the model checkpoint.

File: scripts/llm/finetuning.py
Comment (cuichenx) [reply scripts/llm/finetuning.py:1]: Thanks for the update, sounds good

File: scripts/llm/gpt_train.py [hunk 8]
```diff
@@ -32,34 +33,57 @@
...
+    parser.add_argument("--pp_size", type=int, default=1, help="Pipeline parallel size")
+    parser.add_argument("--ep_size", type=int, default=1, help="Expert parallel size")
+    parser.add_argument("--precision", type=str, default="bf16-mixed", help="Datatype for models and optimizer")
+    parser.add_argument("--devices", type=int, default=1, help="Number of GPUs to use per node")
+    parser.add_argument("--num_nodes", type=int, default=1, help="Number of nodes to use")
+    parser.add_argument("--log_dir", type=str, required=True, help="Folder for logging and checkpoint saving")
+    parser.add_argument("--max_steps", type=int, required=True, help="Number of global batches to process")
+    parser.add_argument("--gbs", type=int, required=True, help="Global Batch Size")
+    parser.add_argument("--mbs", type=int, required=True, help="Micro-batch Size")
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply scripts/llm/gpt_train.py:60]: They should realize if they never provided a dataset that it won't be running on any data

[the remark this answers is not in the document]
Comment (AAnoosheh) [reply scripts/llm/gpt_train.py:60]: Or make it a flag to use mock data mode

File: nemo/collections/llm/api.py (hunk 7 above)
Comment (jenchen13) [reply nemo/collections/llm/api.py]: updated

File: scripts/llm/gpt_train.py [hunk 9]
```diff
@@ -134,16 +178,27 @@ def get_args():
...
+            teacher_model_path=args.teacher_path,
+            data=data,
+            trainer=trainer,
+            log=logger,
+            resume=resume,
+            optim=optim,
+            tokenizer=get_tokenizer(args.tokenizer) if args.tokenizer else None,
+        )
+    else:
+        llm.finetune(
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply scripts/llm/gpt_train.py]: used llm.train

File: nemo/collections/llm/api.py [hunk 10]
```diff
@@ -181,28 +181,33 @@ def pretrain(
 
 @run.cli.entrypoint(namespace="llm")
 def finetune(
-    model: pl.LightningModule,
+    model: Union[pl.LightningModule, str],
```
Comment (AAnoosheh) [inline nemo/collections/llm/api.py:184]: Can we do `AnyPath` (defined in top of this file) in place of `str`

File: scripts/llm/train.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+            data=data,
+            trainer=trainer,
+            log=logger,
+            resume=resume,
+            optim=optim,
+            tokenizer=get_tokenizer(args.tokenizer) if args.tokenizer else None,
+        )
+    else:
+        llm.train(
+            model=args.model_path,
```
Comment (AAnoosheh) [inline scripts/llm/train.py:213]: Now `train()` will need to support model as an `AnyPath` type too

Comment (jenchen13) [reply scripts/llm/train.py]: fixed, i updated train/finetune/pretrain to all support AnyPath too

Comment (AAnoosheh) [reply scripts/llm/train.py]: `_validate_config` gonna complain if the model is str/path

Comment (jenchen13) [reply scripts/llm/train.py]: oh good point, will revert on finetune/pretrain

File: scripts/llm/gpt_train.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jenchen13) [reply scripts/llm/gpt_train.py:72]: done

File: scripts/llm/gpt_train.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+# Suppress lengthy HF warning
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+
+def get_args():
+    """Parse the command line arguments."""
+    parser = ArgumentParser(
+        description="""
+            Script for training GPT models. Supports 4 modes, with different arguments needed in addition to the required arguments:
+            1. Pretrain: no additional arguments required
```
Comment (jenchen13) [inline scripts/llm/gpt_train.py:42]: I finished testing all 4 scenarios and confirmed the initial loss looks good.

**Review by cuichenx — APPROVED** [review]
LGTM

File: scripts/llm/gpt_train.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply scripts/llm/gpt_train.py:72]: Excellent, thanks! Though heads up: I think `len(args.data_paths)` on line 156 & 163 might have issue since the default value will still be `None` so might be better to check if `not args.data_paths` or set its default value to `[]`

**Comment by AAnoosheh** [comment]
LGTM too, after my last thread is resolved :)

File: scripts/llm/gpt_train.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jenchen13) [reply scripts/llm/gpt_train.py:72]: fixed

**Review by AAnoosheh — DISMISSED** [review]

File: scripts/llm/gpt_train.py
[the remark this answers is not in the document]
Comment (jenchen13) [reply scripts/llm/gpt_train.py:1]: thanks, fixed

**Review by cuichenx — APPROVED** [review]

**Review by kevalmorabia97 — APPROVED** [review]

**Comment by jenchen13** [comment]
@soluwalana we discussed offline and agreed to merge this PR first, then rebase #13367 to add additional features
